perf/light.cc

```diff
@@ -0,0 +1,652 @@
+		data.shuffle();
+		state.ResumeTiming();
+
+		struct light_iterator iter;
+		light_iterator_begin(&ht, &iter);
+		Value_t *p = nullptr;
+		while ((p = light_iterator_get_and_next(&ht, &iter)) != nullptr) {
+			benchmark::DoNotOptimize(p->data);
+			processed++;
+		}
+	}
+	state.SetItemsProcessed(processed);
+}
+
+// Fill in hash table, then erase all elements and re-fill it once again
+// (only insertion time is measured).
+BENCHMARK_DEFINE_F(LightFixture, InsertAfterErase)
+(benchmark::State& state)
+{
+	std::size_t processed = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		TupleHolder data(state.range(0));
+		Fill(data.tuples.begin(), data.tuples.end());
+		Erase(data.tuples.begin(), data.tuples.end());
+		data.shuffle();
+		state.ResumeTiming();
+
+		for (const auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_insert(&ht, v.hash, v.raw));
+			processed++;
+		}
+	}
+	state.SetItemsProcessed(processed);
+}
+
+// Fill in hash table, then erase half elements and process lookups
+// (only lookup time is measured).
+BENCHMARK_DEFINE_F(LightFixture, FindAfterErase)
+(benchmark::State& state)
+{
+	std::size_t processed = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		// Double the data size so that after shrink it'll be of the
+		// same size as in other benches.
+		std::size_t data_size = state.range(0) << 1;
+		TupleHolder data(data_size);
+		Fill(data.tuples.begin(), data.tuples.end());
+		Erase(data.tuples.begin(), data.tuples.begin() + (data_size >> 1));
+		data.shuffle();
+		state.ResumeTiming();
+
+		for (const auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_find(&ht, v.hash, v.raw));
+			processed++;
+		}
+	}
+	state.SetItemsProcessed(processed);
+}
+
+// Random insert and random lookup. On each iteration it processes insert and
+// after that - lookup (half of lookups - misses).
+BENCHMARK_DEFINE_F(LightFixture, InsertAndFind)
+(benchmark::State& state)
+{
+	std::size_t processed = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		TupleHolder data(state.range(0));
+		state.ResumeTiming();
+
+		auto lookup_iter = data.tuples.rbegin();
+		for (const auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_insert(&ht, v.hash, v.raw));
+			benchmark::DoNotOptimize(light_find(&ht,
+				lookup_iter->hash, lookup_iter->raw));
+			lookup_iter++;
+			processed += 2;
+		}
+	}
+	state.SetItemsProcessed(processed);
+}
+
+// Delete random values till the hashtable is empty.
+BENCHMARK_DEFINE_F(LightFixture, DeleteRandValue)
+(benchmark::State& state)
+{
+	std::size_t processed = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		TupleHolder data(state.range(0));
+		Fill(data.tuples.begin(), data.tuples.end());
+		data.shuffle();
+		state.ResumeTiming();
+
+		for (const auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_delete_value(&ht, v.hash, v.raw));
+			processed++;
+		}
+	}
+	state.SetItemsProcessed(processed);
+}
+
+BENCHMARK_REGISTER_F(LightFixture, InsertRandValue)->
```

Comment:
Ni (remark for the future): please place `BENCHMARK_REGISTER`s near the corresponding benchmarks, because it becomes hard to understand what the benchmark argument is.

Reply:
Ok, fixed.